freqtrade/optimize/hyperopt/hyperopt.py

```diff
@@ -169,8 +170,20 @@ def get_optuna_asked_points(self, n_points: int, dimensions: dict) -> list[Any]:
             asked.append(self.opt.ask(dimensions))
         return asked
 
+    def check_optuna_asked_points(self, trial: Trial) -> bool:
+        trials_to_consider = trial.study.get_trials(deepcopy=False, states=[TrialState.COMPLETE])
+        # Check whether we already evaluated the sampled `params`.
+        for t in reversed(trials_to_consider):
+            if trial.params == t.params:
+                logger.warning(
+                    f"duplicate trial: Trial {trial.number} has same params as {t.number}"
+                )
+                return True
+        return False
```

Comment:
where would you get the "existing" data from? 
does each `trial` contain the link back to the same `.study`? or is it a new study object each time we call `get_asked_points()`? 

if it's the same study object - doesn't optuna do the "is a duplicate point" detection on it's own? why would it return the same trial multiple times within the same study?

Reply:
running hyperopt for buy sell stoploss roi for 200 epochs result in 3 duplicates warnings.
I'll run now the test for protection only.
And later the both again, but for 1000 epochs.

Reply:
i think you can pretty easily force this by using a very low search-space - like - 1 parameter of ints from 10-20 or so ... it should run out of space latest at epoch 11

Reply:
make sense. I'll test for shorter timerange. 1 month of 5 min? or less?

Reply:
sounds/looks like `.ask()` does increase linearily over epochs, while `.tell()` is neglectible` - and backtesting fluctuates (i suspect based on number of trades made - which is kinda expected)?

How big was the space (that's more relevant than if you used "buy" or "sell" for `--space` ;) )? 
(buy/sell doesn't tell me if there's 2 or 200 parameters per space - so an overall "there's 200 parameters" is more telling than WHICH space we optimized - that's only relevant for the strategy - for optuna, it's all the same).

Anyway what's the conclusion? it's a trade off either way?

Reply:
oh i have 0 doubt that it can be simplified :laughing: 
it's more a question of "how much" and which features still makes sense considering that optuna seems to be able to get new points faster (this might depend on the model??)

Reply:
Test results after adding 5 ask retries:
space: buy sell stoploss roi / epochs: 1000 - duplicates: 0
<img width="940" alt="Screenshot 2025-05-28 at 08 07 31" src="https://github.com/user-attachments/assets/5af54171-529e-406d-9b3d-92ff24d5a7e9" />
space: protection / epochs: 1000 - duplicates: testing

Reply:
1s is ... low enough - not worth additional logics (like fetching more than we need).

Reply:
space: protection / epochs: 1000 - duplicates: 0
<img width="1345" alt="Screenshot 2025-05-28 at 09 49 37" src="https://github.com/user-attachments/assets/e1366b18-af76-472f-8dc3-ede9c2759c18" />
All good now.
Please let me know what to do with is_random. Now it is only False

Reply:
Something is still wrong. I need to check more:
1st batch:
get_asked_points output: `[{'stoploss': -0.031}, {'stoploss': -0.05600000000000001}, {'stoploss': -0.015}, {'stoploss': -0.091}, {'stoploss': -0.063}, {'stoploss': -0.033}, {'stoploss': -0.015}, {'stoploss': -0.017}, {'stoploss': -0.05}, {'stoploss': -0.068}, {'stoploss': -0.038000000000000006}, {'stoploss': -0.05600000000000001}, {'stoploss': -0.048}, {'stoploss': -0.02600000000000001}]`
2nd batch:
asked_non_tried: `[{'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}, {'stoploss': -0.1}]`

Reply:
I think I discovered the issue with same parameters in the same batch: this is happening with only with Autosampler and only when one space is optimized. Maybe this is the way this sampler is warming up and learning.

Now I'm testing with default NSGAIIISampler and I do not see so far duplicates in batch and in history.
I'll post results after test finish.

Reply:
Looks fine now, no params duplicates for same batch or already used.

The warning message ""duplicate params detected. Please check if search space is not too small!"" will appear every batch in case space is exhausted. Maybe LoggingMixin with log_once shall be used.

Reply:
maybe `population_size` should then be set to `INITIAL_POINTS` ? 

That's how skopt did ... and would allow users to control it (to some degree).
Could obviously also be made into a config parameter ... but i'm not sure 50 is ideal for all scenarios - we've seen 30 to be a more reasonable case from history...

Reply:
Maybe users can use generate_estimator to customize parameters…
And, yes, i’ll test these days with 30, 50, 70 and post results

Reply:
hm if `population_size` is not a generic parameter across (textbased-supported) generators - absolutely, this would make it a "docs only" topic.

Reply:
population size is not generic parameter. TPESampler has n_startup_trials=10, NSGAIIISampler has pupulation_size.

Our default sampler is NSGAIIISampler and make sense to tune it now. 
Any suggestion for some strategies to be tested with population_size 30, 50, 70?

Reply:
not really - it'll have to be smallish epoch sizes i guess - most likely with huge search spaces (where there's tons of local minima).

if the X points picked initially include very good starting points, you can still get good results - so that's not ... "perfect science" unfortunately

Reply:
30 is a better value, at least for the strategy I tested:
INITIAL_POINTS: 30 / TTR: 3:32:00 / Profit: 18,535.70  / Winrate: 63.5 / Mem usage: 73GB
INITIAL_POINTS: 50 / TTR: 3:15:00 / Profit: 11,417.62 / Winrate: 59.0 / Mem usage: 73GB
INITIAL_POINTS: 70 / TTR:  / Profit: 12,907.42 / Winrate: 60.1 / Mem usage: 73GB

Reply:
well - so we used the "best" value already :laughing: 
which is funny - as it's not guaranteed / given that the same value would be good for a completely new library.

Thanks for your work and efforts here! i'll take a look - and will try to include it in the release i plan on doing this weekend.

Reply:
no pb.
Interesting indeed. 
I started a test with values between 10 and 100. If I'll find any other value significantly better than 30, I'll let you know. Also when time permits, I'll download one strategy with many more parameters to see how it goes.
This does not stop your work, you can close this topic whenever you want.